### PR TITLE
CCIP-4447 Promwrapper for OCR3 plugins and factories 

### PR DESCRIPTION
**Pull request by mateusz-sekara**
Implementing a very basic monitoring layer for OCR3 Reporting Plugins to support similar behavior to what we have for OCR2 https://github.com/smartcontractkit/chainlink/blob/develop/core/services/ocr2/plugins/promwrapper/plugin.go

Compared to the OCR2 implementation some labels like `chainType`, `configDigest` and `oracleID` were removed. This is done to reduce the cardinality of the metrics published to Prometheus. Intentionally, starting with a fundamental set of metrics which can be extended in the future if needed.

These metrics will help us answer the following observability questions:
* what is the error rate for different OCR3 phases
* what is the throughput per plugin (per OCR3 phase)
* what are the latencies per OCR3 phase 
* how many reports are generated by plugins
* which oracles/plugins are accepting/transmitting reports

### CODE REVIEW

File: core/services/ocr3/promwrapper/types.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	observation         functionType = "observation"
+	validateObservation functionType = "validateObservation"
+	outcome             functionType = "outcome"
+	reports             functionType = "reports"
+	shouldAccept        functionType = "shouldAccept"
+	shouldTransmit      functionType = "shouldTransmit"
+)
+
+var (
+	buckets = []float64{
```
Comment (emate) [inline core/services/ocr3/promwrapper/types.go:23]: What's the use case for this number of buckets?

File: core/services/ocr3/promwrapper/types.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+		float64(500 * time.Millisecond),
+		float64(700 * time.Millisecond),
+		float64(time.Second),
+		float64(2 * time.Second),
+		float64(5 * time.Second),
+		float64(10 * time.Second),
+	}
+
+	promOCR3ReportsGenerated = promauto.NewGaugeVec(
+		prometheus.GaugeOpts{
```
Comment (emate) [inline core/services/ocr3/promwrapper/types.go]: Wouldn't `counter` be better here since it will only increment?

**Review by emate — APPROVED** [review]

**Review by makramkd — APPROVED** [review]

**Review by samsondav — APPROVED** [review]
Nice! Can you add it to the LLO plugin in core/services/llo/delegate.go?

File: core/services/ocr3/promwrapper/plugin.go [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+	return withObservedExecution(p, outcome, func() (ocr3types.Outcome, error) {
+		return p.ReportingPlugin.Outcome(ctx, outctx, query, aos)
+	})
+}
+
+func (p *reportingPlugin[RI]) Reports(ctx context.Context, seqNr uint64, outcome ocr3types.Outcome) ([]ocr3types.ReportPlus[RI], error) {
+	result, err := withObservedExecution(p, reports, func() ([]ocr3types.ReportPlus[RI], error) {
+		return p.ReportingPlugin.Reports(ctx, seqNr, outcome)
+	})
+	p.trackReports(reports, len(result))
```
Comment (0xnogo) [inline core/services/ocr3/promwrapper/plugin.go:71]: Does it mean that we are tracking report even when we have an error?

File: core/services/ocr3/promwrapper/types.go [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+	observation         functionType = "observation"
+	validateObservation functionType = "validateObservation"
+	outcome             functionType = "outcome"
+	reports             functionType = "reports"
+	shouldAccept        functionType = "shouldAccept"
+	shouldTransmit      functionType = "shouldTransmit"
+)
+
+var (
+	buckets = []float64{
```
Comment (0xnogo) [inline core/services/ocr3/promwrapper/types.go:23]: Are these buckets applicable for all chains? Solana might be significantly faster than Ethereum. I wonder if we should have buckets per chain.

**Review by 0xnogo — APPROVED** [review]
LGTM, i had few minors

File: core/services/ocr3/promwrapper/plugin.go (hunk 3 above)
Comment (mateusz-sekara) [reply core/services/ocr3/promwrapper/plugin.go:71]: reports should be nil or empty in that case

File: core/services/ocr3/promwrapper/types.go (hunk 4 above)
Comment (mateusz-sekara) [reply core/services/ocr3/promwrapper/types.go:23]: For CCIP it shouldn't be a big deal. Most of the work we do in OCR is reading from the database (ChainReader), doing some data transformation, reaching 3rd party APIs etc. There is still some communication with the chain involved but it shouldn't be that relevant. However, we might come back to that after seeing different values on non-evm chains

Comment (mateusz-sekara) [reply core/services/ocr3/promwrapper/types.go:23]: To answer your question: yes, buckets are applicable for all chains

**Pull request merged**